### PR TITLE
Add Finch.Error exception

### PR DESCRIPTION
**Pull request by lukebakken**
Fixes https://github.com/keathley/finch/issues/142

### CODE REVIEW

File: lib/finch.ex [hunk 1]
```diff
@@ -224,7 +224,7 @@ defmodule Finch do
 
   """
   @spec stream(Request.t(), name(), acc, stream(acc), keyword) ::
-          {:ok, acc} | {:error, Mint.Types.error()}
+          {:ok, acc} | {:error, Mint.Types.error()} | {:error, Finch.Error.t()}
```
Comment (wojtekmach) [inline lib/finch.ex:227]: Maybe we should use this opportunity to also solve: https://github.com/keathley/finch/issues/54#issuecomment-645381816?

```suggestion
          {:ok, acc} | {:error, Exception.t()}
```

File: lib/finch/error.ex [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+  A `Finch.Error` struct is an exception, so it can be raised as any other exception.
+
+  ## Message representation
+
+  If you want to convert an error reason to a human-friendly message (for example
+  for using in logs), you can use `Exception.message/1`:
+
+      iex> {:error, %Finch.Error{} = error} = request...
+      iex> Exception.message(error)
+
```
Comment (wojtekmach) [inline lib/finch/error.ex:17]: This is general information about exceptions and not specific to this particular one so I'd remove it.

```suggestion
```

File: lib/finch/error.ex [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+    %__MODULE__{reason: reason}
+  end
+
+  @impl true
+  def message(%__MODULE__{reason: reason}) when is_atom(reason) do
+    "error: #{reason}"
+  end
+
+  @impl true
+  def message(%__MODULE__{reason: reason}) do
```
Comment (wojtekmach) [inline lib/finch/error.ex:35]: this clause is unused and I'd consider removing it.

File: lib/finch/error.ex [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+
+  @impl true
+  def exception(reason) do
+    %__MODULE__{reason: reason}
+  end
+
+  @impl true
+  def message(%__MODULE__{reason: reason}) when is_atom(reason) do
+    "error: #{reason}"
+  end
```
Comment (wojtekmach) [inline lib/finch/error.ex:32]:
```suggestion
  @impl true
  def exception(reason) when is_atom(atom) do
    %__MODULE__{reason: reason}
  end

  @impl true
  def message(%__MODULE__{reason: reason}) do
    "error: #{reason}"
  end
```

File: test/finch/http2/pool_test.exs [hunk 5]
```diff
@@ -92,7 +92,8 @@ defmodule Finch.HTTP2.PoolTest do
     :timer.sleep(10)
 
     # We can't send any more requests since the connection is closed for writing.
-    assert {:error, %{reason: :read_only}} = request(pool, req, [])
+    assert {:error, %Finch.Error{reason: :read_only}=error_read_only} = request(pool, req, [])
+    assert Exception.message(error_read_only)
```
Comment (wojtekmach) [inline test/finch/http2/pool_test.exs:96]: I'd keep it similar to other tests you added below. This means we'd no longer be testing our message/1 implementation but that's imho totally fine :)

```suggestion
    assert {:error, %Finch.Error{reason: :read_only}} = request(pool, req, [])
```

File: test/finch/http2/pool_test.exs [hunk 6]
```diff
@@ -107,7 +108,8 @@ defmodule Finch.HTTP2.PoolTest do
     Process.sleep(50)
 
     # If we try to make a request now that the server shut down, we get an error.
-    assert {:error, %{reason: :disconnected}} = request(pool, req, [])
+    assert {:error, %Finch.Error{reason: :disconnected}=error_disconnected} = request(pool, req, [])
+    assert Exception.message(error_disconnected)
```
Comment (wojtekmach) [inline test/finch/http2/pool_test.exs:112]:
```suggestion
    assert {:error, %Finch.Error{reason: :disconnected}} = request(pool, req, [])
```

File: lib/finch/error.ex [hunk 7]
```diff
@@ -0,0 +1,38 @@
...
+  defexception [:reason]
+
+  @impl true
+  def exception(reason) do
+    %__MODULE__{reason: reason}
+  end
+
+  @impl true
+  def message(%__MODULE__{reason: reason}) when is_atom(reason) do
+    "error: #{reason}"
```
Comment (wojtekmach) [inline lib/finch/error.ex:31]: nitpick: this is going to be logged as:

```
** (Finch.Error) error: connection_closed
```

and "just" error feels a bit redundant to me but it's pretty subjective. some other ideas:

1. `** (Finch.Error) connection_closed`
2. `** (Finch.Error) HTTP/2 error: connection_closed`
3. `** (Finch.HTTP2Error) connection_closed`

**Review by wojtekmach — COMMENTED** [review]
Just some unsolicited feedback :)

**Comment by lukebakken** [comment]
@wojtekmach great, thanks. I figured there would be decisions I made that needed adjusting. I'll get the changes done.

File: lib/finch.ex (hunk 1 above)
Comment (lukebakken) [reply lib/finch.ex:227]: Let me know what else would need to be done to address #54.

File: lib/finch/error.ex (hunk 7 above)
Comment (lukebakken) [reply lib/finch/error.ex:31]: Great, removing the string `error: `

**Comment by lukebakken** [comment]
@wojtekmach all set for round two. GitHub won't let me "re-request review" 🤷

**Comment by wojtekmach** [comment]
LGTM!

**Comment by lukebakken** [comment]
Thanks @wojtekmach - I appreciate the very fast PR reviews.

**Review by keathley — APPROVED** [review]

**Pull request merged**